validate.go

```diff
@@ -432,22 +433,37 @@ func (v *Validator) HexColor(key, value string, message ...string) {
 // Len sets the minimum and maximum length for a string.
 //
 // A maximum of 0 indicates there is no upper limit.
-func (v *Validator) Len(key, value string, min, max int, message ...string) {
+func (v *Validator) Len(key string, value interface{}, min, max int, message ...string) {
 	msg := getMessage(message, "")
-
-	switch {
-	case len(value) < min:
-		if msg != "" {
-			v.Append(key, msg)
-		} else {
-			v.Append(key, fmt.Sprintf(MessageLenLonger, min))
+	t := reflect.TypeOf(value)
+	k := t.Kind()
+	val := reflect.ValueOf(value)
+
+	switch k {
+	case reflect.String,
+		reflect.Slice,
+		reflect.Map,
+		reflect.Array,
+		reflect.Chan:
+
+		if val.Len() < min {
+			if msg != "" {
+				v.Append(key, msg)
+			} else {
+				v.Append(key, fmt.Sprintf(MessageLenLonger, min))
+			}
 		}
-	case max > 0 && len(value) > max:
-		if msg != "" {
-			v.Append(key, msg)
-		} else {
-			v.Append(key, fmt.Sprintf(MessageLenShorter, max))
+
+		if max > 0 && val.Len() > max {
+			if msg != "" {
+				v.Append(key, msg)
+			} else {
+				v.Append(key, fmt.Sprintf(MessageLenShorter, max))
+			}
 		}
+
+	default:
+		v.Append(key, fmt.Sprintf(MessageLenInvalidType, k.String()))
```

Comment:
This is not a validation error; what is the user supposed to do with this error? It's a programming error.

In Required() it panics: https://github.com/Teamwork/validate/blob/227d00d848e48302e79731e7cefb16dbd35f8854/validate.go#L252

Reply:
I disagree, for example if the user passes a int value and the programmer does not validate the type (string for example), doesn't seem that it should result in a panic/500 error.

It is weird though that he receives a 400 for length, but seems better than panic.

@ripexz Any thoughts on this?

Reply:
The user doesn't pass any values though; it's scanned in to a `string`, `int`, or something else, and that's passed to this library. The user doesn't control the type that gets passed to this function, the programmer does.

The entire point of this library is to avoid all this `interface{}` muckery in the first place and rely on Go's typing and compiler errors, with `Required()` being a special case. Hence my "do we really need this?"-comment, because this will chuck all the compiler safety out the window.

Reply:
I mean, the question here is "what concrete problem are you trying to solve?" because right now this patch is quite a paradigm shift. Perhaps there is a good reason for this, but it's also possible it can be solved in a much better way.

Reply:
This being an open source library, it seems it should be agnostic on the use cases. For that reason, panicking it's always a pain to handle, but I can go with it if you have a strong opinion on that.

Reply:
You don't need to "handle" any panics for this though, this case is *expected* to panic and stop all code, as it's clearly a programming error. That's pretty much the intended use. Some people overuse panics, but that doesn't mean there aren't valid use cases.

> This being an open source library, it seems it should be agnostic on the use cases.

Well, the new code has some serious downsides the previous code didn't have, aside from making it inconsistent. Just because it handles more use cases doesn't mean it's "better". Code is designed to solve problems, hence my question: what problems does this solve?

I would strongly prefer to avoid `interface{}` altogether if possible, and instead opt for a new `StringCount(key string, value []string)` function for example. But I don't know if that's an appropriate solution if I don't know what problem this is intended to solve.